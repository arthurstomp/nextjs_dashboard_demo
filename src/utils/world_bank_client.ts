import { request, gql } from 'graphql-request'

export const buildQuery = (countryCode: string, indicatorCode: string, startingYear: string, endingYear: string) => (
  gql`
  {
    ${countryCode}: getWorldBankIndicators(country_iso_code: "${countryCode}", indicator_code: "${indicatorCode}", starting_year: "${startingYear}", ending_year: "${endingYear}") {
      year
      value
    }
  }
  `
)

function compareYear(a: any, b: any) {
  const aYearInt = parseInt(a.year)
  const bYearInt = parseInt(b.year)

  if (aYearInt > bYearInt) {
    return 1
  }

  if (aYearInt < bYearInt) {
    return -1
  }
  return 0
}

export type ProtoDataset = {
  labels: string[]
  data: number[]
}

function resultToProtoDataset(result: any[]) {
  let buf: ProtoDataset = {
    labels: [],
    data: []
  }
  result.forEach((res: any) => {
    buf.labels.push(res.year)
    buf.data.push(res.value)
  })

  return buf
}

export const fetcher = async ([countryCode, indicatorCode, startingYear, endingYear]: string[]) => {
  console.log('startingYear', startingYear)
  console.log('endingYear', endingYear)
  let res = await request('/api/graphql', buildQuery(countryCode, indicatorCode, startingYear, endingYear))
  res = (res as any)[countryCode]
  res = (res as any[]).sort(compareYear)
  return resultToProtoDataset(res as any[])
}
