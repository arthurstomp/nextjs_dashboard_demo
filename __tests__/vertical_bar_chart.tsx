import React from 'react'
import VerticalBarChart from "@/components/vertical_bar_chart";
import { render } from "./test-utils";
import '@testing-library/jest-dom'
import { SWRConfig } from "swr";

global.ResizeObserver = jest.fn().mockImplementation(() => ({
    observe: jest.fn(),
    unobserve: jest.fn(),
    disconnect: jest.fn(),
}))


jest.mock('@/utils/world_bank_client', () => {
  return({
    __esModule: true,
    fetcher: jest.fn(() => {
      return {
        labels: ['2021'],
        data: []
      }
    })
  })
})

import { fetcher } from '@/utils/world_bank_client'

describe("VerticalBarChart", () => {
  it("matches snapshot", async () => {
    const { baseElement } = render(
      <SWRConfig value={{ provider: () => new Map() }}>
        <VerticalBarChart ></VerticalBarChart>
      </SWRConfig>
    )

    expect(baseElement).toMatchSnapshot()
  })

  it('calls for world_bank_clinet', async () => {
    const { findByTestId } = render(
      <SWRConfig value={{ dedupingInterval: 0, provider: () => new Map() }}>
        <VerticalBarChart ></VerticalBarChart>
      </SWRConfig>
    )

    await findByTestId('chart')
    expect(fetcher).toHaveBeenCalled()
  })
})
