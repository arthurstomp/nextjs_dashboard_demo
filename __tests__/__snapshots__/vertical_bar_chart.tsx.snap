// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VerticalBarChart matches snapshot 1`] = `
<body>
  <div>
    <div
      data-testid="chart"
    >
      <canvas
        height="0"
        role="img"
        style="display: block; box-sizing: border-box; height: 0px; width: 0px;"
        width="0"
      />
    </div>
  </div>
</body>
`;
